### PR TITLE
ADuCM360_demo_cn0359: Initial commit

### PR DESCRIPTION
**Pull request by rbolboac**
The cn0359 device is a completely self-contained, microprocessor
controlled, highly accurate conductivity measurement system ideal
for measuring the ionic content of liquids, water quality
analysis, industrial quality control, and chemical analysis.

A carefully selected combination of precision signal
conditioning components yields an accuracy of better than 0.3%
over a conductivity range of 0.1 μS to 10 S (10 MΩ to 0.1 Ω)
with no calibration requirements.

Automatic detection is provided for either 100 Ω or 1000 Ω
platinum (Pt) resistance temperature devices (RTDs), allowing
the conductivity measurement to be referenced to 25°C.

The system accommodates 2- or 4-wire conductivity cells,
and 2-, 3-, or 4-wire RTDs for added accuracy and flexibility.

Signed-off-by: Ramona Bolboaca <ramona.bolboaca@analog.com>

This is an older project (author unknown). My changes are in lcd.cpp file and small changes in other files (adding copyright header and fixed astyle and compilation errors).
New changes have been added to support ADALM-UARTJTAG connector for serial communication instead of RS-485 interface.

### CODE REVIEW

**Comment by rbolboac** [comment]
- fixed astyle and added copyright header to all files

File: projects/ADuCM360_demo_cn0359/src/hal/drivers/lcd.cpp [hunk 1]
```diff
@@ -0,0 +1,591 @@
...
+void static lcd_framebuffer_flush(void)
+{
+	timer t;
+	t.time = FB_FLUSH_DELAY;
+	t.timer_app.fun = send_spi_image;
+	t.timer_app.argc = 0;
+	t.timer_app.argv = new char*;
+
+	update_timer(t);
+
```
Comment (CiprianRegus) [inline projects/ADuCM360_demo_cn0359/src/hal/drivers/lcd.cpp:368]: Extra line.

File: projects/ADuCM360_demo_cn0359/src/hal/drivers/lcd.cpp [hunk 2]
```diff
@@ -0,0 +1,591 @@
...
+{
+	lcd_write_control(0xA0); // ADC select
+	lcd_write_control(0xAE); // Display OFF
+	lcd_write_control(0xC8); // COM direction scan
+	lcd_write_control(0xA2); // LCD bias set
+	lcd_write_control(0x2F); // Power Control set
+	lcd_write_control(0x21); // Resistor Ratio Set
+
+	flash_file * p_flash_file;
+	decltype(flash_file::brightness) brightness;
```
Comment (CiprianRegus) [inline projects/ADuCM360_demo_cn0359/src/hal/drivers/lcd.cpp:512]: nitpick: I think these should be declared at the beginning of the function.

File: projects/ADuCM360_demo_cn0359/src/hal/drivers/lcd.cpp [hunk 3]
```diff
@@ -0,0 +1,591 @@
...
+			errno = EINVAL;
+			return -1;
+		} else {
+			tty_cursor = sizeof(lcd_tty_buffer) + offset;
+		}
+		break;
+	default:
+		errno = ENXIO;
+		return -1;
+		break;
```
Comment (CiprianRegus) [inline projects/ADuCM360_demo_cn0359/src/hal/drivers/lcd.cpp:496]: Useless break.

File: projects/ADuCM360_demo_cn0359/src/hal/drivers/lcd.cpp [hunk 4]
```diff
@@ -0,0 +1,591 @@
...
+off_t lcd_tty_lseek(off_t offset, int whence)
+{
+	switch (whence) {
+	case SEEK_SET:
+		if (offset > sizeof(lcd_tty_buffer) || offset < 0) {
+			errno = EINVAL;
+			return -1;
+		} else {
+			tty_cursor = offset;
+		}
```
Comment (CiprianRegus) [inline projects/ADuCM360_demo_cn0359/src/hal/drivers/lcd.cpp:473]: Useless else (same for the SEEK_CUR and SEEK_END cases).

File: projects/ADuCM360_demo_cn0359/src/hal/drivers/lcd.cpp [hunk 5]
```diff
@@ -0,0 +1,591 @@
...
+		// 32pixel display / 8 pixels per page = 4 pages
+		lcd_write_control(page); // send page address
+
+		// Sets the most significant 4 bits of the display RAM column address.
+		lcd_write_control(0x10); // column address upper 4 bits + 0x10
+		// Sets the least significant 4 bits of the display RAM column address.
+		lcd_write_control(0x00); // column address lower 4 bits + 0x00
+		for (unsigned int j = 0; j < 128; j++) {
+			// 128 columns wide
+			// todo
```
Comment (CiprianRegus) [inline projects/ADuCM360_demo_cn0359/src/hal/drivers/lcd.cpp:540]: nitpick: is this needed?

File: projects/ADuCM360_demo_cn0359/src/hal/drivers/lcd.cpp [hunk 6]
```diff
@@ -0,0 +1,591 @@
...
+}
+
+void static lcd_write_image(const unsigned char data)
+{
+	GP0SET_SET4_BBA = true;
+	pADI_SPI1->SPITX = data;
+	while ((pADI_SPI1->SPISTA & SPI1STA_TXFSTA_MSK) != SPI1STA_TXFSTA_EMPTY)
+		;
+}
+
```
Comment (CiprianRegus) [inline projects/ADuCM360_demo_cn0359/src/hal/drivers/lcd.cpp:337]: Extra line.

File: projects/ADuCM360_demo_cn0359/src/hal/drivers/lcd.cpp [hunk 7]
```diff
@@ -0,0 +1,591 @@
...
+			// 128 columns wide
+			// todo
+			lcd_write_image(framebuffer_memory[i][j]); // send picture data
+		}
+		page++; // after 128 columns, go to next page
+	}
+	// lcd_write_control(0xAF); // Display on
+	return 0;
+}
+void static lcd_framebuffer_flush(void)
```
Comment (CiprianRegus) [inline projects/ADuCM360_demo_cn0359/src/hal/drivers/lcd.cpp:359]: There's a new line needed here.

**Review by CiprianRegus — COMMENTED** [review]
Looks good, only a few nitpicks from my side.

**Comment by CiprianRegus** [comment]
Also, the commit title has a typo (DuCM360).

**Review by amiclaus — APPROVED** [review]
lgtm, after @CiprianRegus remarks are addressed.

**Comment by rbolboac** [comment]
- Implemented review changes
- New changes have been added to support ADALM-UARTJTAG connector for serial communication instead of RS-485 interface.

**Comment by bbushey** [comment]
Looking good Ramona, please make sure that when you are complete here and everything is merged that you work with Allan to update the wiki page.  It looks like there is some significant changes with the setup and how to communicate with the hardware via the ADALM-UARTJTAG that he should document for customers.

**Review by CiprianRegus — APPROVED** [review]

**Pull request merged**